api/src/main/java/jakarta/json/bind/annotation/JsonbPolymorphicType.java

```diff
@@ -0,0 +1,163 @@
+ * <br>
+ * This annotation is required to be present on the most top of all the classes,
+ * polymorphism should be applied to.
+ * <pre>{@code
+ * // Example
+ * @JsonbPolymorphicType(keyName = "@key")
+ * interface Animal {}
+ *
+ * class Dog implements Animal {}
+ * class Cat implements Animal {}
+ * }</pre>
+ * This annotation is closely tight with {@link JsonbSubtype}. It is recommended to use
+ * {@link JsonbSubtype} annotations to specify all the possible classes and their aliases.
+ * <br>
+ */
+@JsonbAnnotation
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE})
+public @interface JsonbPolymorphicType {
+
+    /**
+     * Key name used for keeping polymorphic information when {@link Format#PROPERTY} is chosen.
+     *
+     * @return key name
+     */
+    String keyName() default "";
+
+    /**
+     * Specification of how the polymorphic information will be stored in the resulting JSON.
+     *
+     * @return polymorphic information storage type
+     */
+    Format format() default Format.PROPERTY;
+
+    /**
+     * Whether exact class names should be processed if no alias is specified for processed class.
+     * <br>
+     * Default value is false.
+     *
+     * @return classes without alias should be processed
+     */
+    boolean classNames() default false;
+
+    /**
+     * Whitelisted package names. This option is ignored if {@link JsonbPolymorphicType#classNames()}
+     * option is set to false.
+     * <br>
+     * Only classes contained in the selected packages will be serialized/deserialized.
+     * Classes with specified alias are not validated.
+     * <br>
+     * It is strongly recommended having whitelist set up if classes without alias should be processed.
+     * <br>
+     * When no whitelist is specified, all classes without alias are allowed.
+     *
+     * @return package whitelist
+     */
+    String[] whitelist() default "";
+
+    /**
+     * Format under which serialized polymorphic type should be handled.
+     * <br>
+     * Default polymorphic format is {@link Format#PROPERTY}.
+     */
+    enum Format {
+
+        /**
+         * Serialized object will be wrapped with another object.
+         * This wrapping object will contain only one property with the name of the
+         * polymorphic information (alias/class name) and a value assigned to this property is an actual
+         * serialized object.
+         * <br>
+         * This format ignores specified {@link JsonbPolymorphicType#keyName()}.
+         * <pre>{@code
+         * // Example
+         * @JsonbPolymorphicType(format = Format.WRAPPING_OBJECT)
+         * @JsonbSubtype(alias = "dog", type = Dog.class)
+         * interface Animal {}
+         *
+         * class Dog implements Animal {
+         *     public String isDog = true;
+         * }
+         *
+         * //Resulting json after dog instance serialization
+         * {"dog":{"isDog":true}}
+         * }</pre>
+         */
+        WRAPPING_OBJECT,
```

Comment:
if we support a wrapping_object i'd be for `{type:xxx, value:{...}}`, not that which is quite specific

Reply:
Not necessary. You would have redundant field, since it can be merged into one like I proposed.
Instead of: `{ "type":"someAlias", "value":{...}}` you can have `{"someAlias":{...}}`

Reply:
it is equivalent in terms of information but not at all in terms of usage and integrations.
it becomes very consumer specific so not sure we should have it as an enum.
starts to looks like we should reuse (de)serializer instead of trying to cover all user cases which is sure we'll fail to do so either we just support property or we use a real SPI but I don't see this enum solving the issue properly from what I saw in past apps.

Reply:
Alright, do you think it would be better to just have only one approach currently supported? Such as the PROPERTY format? Do you think I should drop the rest for now?

Reply:
The reason why I chose to propose all 3 formats, is actually that Jackson does support those and if we want to be interoperable with Jackson, this would ensure it.

Reply:
Yeah, think having a single strategy is fine for now.
About jackson interop: who cares, it is not interop with json.io other polymorphism solutions anyway ;).

Reply:
On one hand this notation could be solved by ugly holder object:
```
class PoorMansPolymorphism {
   public Cat cat;
   public Dog dog;
```

On the other hand this is very common style in everything backed by Go, as that's exactly how they must solve polymorphism, and it would be good if JSON-B offered more in that regard.

Reply:
To be fair, I was thinking a bit about keeping the different styles, I am for to keep it as proposed in the end. I don't see any reason currently why not to have it.

Reply:
@pdudits Would you like me to add some other style in regards of your comment? :-)

Reply:
@Verdent can you add a keyed wrapper too (`{value:{dog:{}}}`), but also a keyed array wrapper, a merge strategy (`{common1:...,dog:{specificForDog:....}}`), a merge wrapper, a configurable merge strategy (so keyName becomes `formatOptions=...` etc...this is the point of not having any format, there are too much ways to represent it and all are valid and used. This is also why we must ensure (de)serializers work to handle that properly, to let user read any incoming json (they dont control) and output any json their contract enforce. So at the end, this feature can only be a default opiniated version so I strongly think formats are not worth it. To be useful it must be a SPI which will end up being the (de)serialization one we already have.

Reply:
@rmannibucau No one forces you to use any of these formats if you don't want to. As you have mentioned in the comment below, interoperability is important and since this is what one of the most widely supported frameworks uses, I think it is important to support it.

Reply:
> what one of the most widely supported frameworks uses

No, it is only in java land, JSON usage is way more than that and just to illustrate it, the interop with javascript breaks that and is way more than jsonb+jackson.

I agree "No one forces you to use any of these formats" but why do you add format you use and not the format I use? In other words why do you make the spec more verbose and complex for some particular cases and not others? One quality of JSON-B is to be simple, I'd love to avoid to make it a new jackson with all its complexity which makes it hard for newcomers and ambiguous to implement a feature (we already have serializers/adapters which are not far).

Reply:
To fully quantify an answer one would need a study of how common a particular style of JSON polymorphic data structure is. I'm not aware of one, neither I have capacity to conduct one.

@rmannibucau do you have some data that might help us decide here?

I personally recognize the first two formats as quite common, and it would be good to support both, for sake of user experience. I wouldn't miss `wrapping_array` at all.

But I also agree that this is not enough for all the cases, and having serializers and adapters work for polymorphic types would be great, once we have solution for recursion trap hidden inside that.

Reply:
nothing strong enough, just know in several apps I saw ~8-9 formats and all were supposed to be "generic" so fit this format definition but as a generic lib I wouldn't use most of them, only type injection "inline" and type+value wrapping makes sense as default IMHO.

Reply:
@pdudits Yes I am alright with letting `wrapping_array` go too :-) Do you agree with keeping `property` and `wrapping_object` the same way they are proposed?

Reply:
think that if we go with some built-in we should probably have "injected type" (`{@type:..., otherAttr:...}`) and "generic wrapper" (`{type:...,value:{}}`), others _seems_ less common. The "type as key" option does not seem used a lot.

@Verdent any idea how far the impl will be from a serializer/deserializer pair? My suspicion is that it will be a "codec" (`JsonbSomeCleverNaming<T> implement JsonbTypeDeserializer<T>, JsonbTypeSerializer<T>`). if so I'd replace format with `Class< extends JsonbSomeCleverNaming<?>>` with a `interface JsonbPolymorphismFormats { interface PropertyTypeInjection {} interface WrappingObject {}}` markers to let the user customize it at need on the interface/parent directly and reuse all we already have which matches this need.

Enables any number of format and to extend it at will in the future if we realize some are more common than we think like the wrapping_object of the javadoc.

Reply:
The way impl will choose to implement this, is completely up to the implementation. Yasson will not go with deserializers/serializers in this matter at all. Having it done the way you are proposing is the same as having serializers/deserializers and registered them to the jsonb. This class, you are having, is just a shortcut for having them both implemented directly. If user wants to do it with serializers/deserializers so be it, but in that case it is not needed to do that over these annotations.